### PR TITLE
For code review

### PR DESCRIPTION
**Pull request by levb**

### CODE REVIEW

File: server/plugin.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		} else if textNode, ok := node.(*markdown.Text); ok {
+			startPos, endPos = textNode.Range.Position+offset, textNode.Range.End+offset
+			origText = postText[startPos:endPos]
+			if textNode.Text != origText {
+				mlog.Error(fmt.Sprintf("Markdown text did not match range text, '%s' != '%s'", textNode.Text, origText))
+				return true
+			}
+		}
+
+		if origText != "" {
```
Comment (levb) [inline server/plugin.go]: recommend either `if origText == "" { return true` or refactor into a function

File: server/plugin.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	p.links.Store(links)
+	return nil
+}
+
+// MessageWillBePosted is invoked when a message is posted by a user before it is committed
+// to the database.
+func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*model.Post, string) {
+	links := p.links.Load().([]*AutoLinker)
+	postText := post.Message
+	offset := 0
```
Comment (levb) [inline server/plugin.go]: nit: for extra clarity, `int(0)` since it can be negative, in theory anyway

File: server/autolinker.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	}, nil
+}
+
+// Replace will subsitute the regex's with the supplied links
+func (l *AutoLinker) Replace(message string) string {
+	if l.pattern == nil {
+		return message
+	}
+
+	// beacuse MMDisableNonWordPrefix DisableNonWordSuffix are greedy then
```
Comment (levb) [inline server/autolinker.go]: nit: maybe worth making the result of the first ReplaceAll() visible to the tests, and add a test that explicitly illustrates how this works

File: server/plugin.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+				team, tErr := p.API.GetTeam(channel.TeamId)
+				if tErr != nil {
+					mlog.Error(tErr.Error())
+					return false
+				}
+				teamName = team.Name
+			}
+
+			for _, l := range links {
+				if len(l.link.Scope) == 0 {
```
Comment (levb) [inline server/plugin.go]: nit: combine the if clauses

File: server/autolinker.go (hunk 3 above)
Comment (jfrerich) [inline server/autolinker.go]: nit: s/beacuse/ because/

File: README.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  - Pattern: `(?i)(ticket )(?P<ticket_id>.+)(:)(?P<ticket_info>.*)`, or if the ticket_id is a number, then `(?i)(ticket )(?P<ticket_id>\d+)(:)(?P<ticket_info>.*)`
+  - Template: `[Ticket ${ticket_id}: ${ticket_info}](https://github.com/mattermost/mattermost-server/issues/${ticket_id})`
+
+2. Autolinking a link to a GitHub PR to a link with format "pr-repo-id". Use:
+  - Pattern: `https://github\\.com/mattermost/(?P<repo>.+)/pull/(?P<id>\\d+)`
+  - Template: `[pr-${repo}-${id}](https://github.com/mattermost/${repo}/pull/${id})`
+
+3. Using autolinking to create group mentions. Use:
+  - Pattern: `@customgroup*`
+  - Template: `[@customgroup]( \\* @user1 @user2 @user3 \\* )`
```
Comment (jfrerich) [inline README.md]: Add some examples which make use of DisableNonWordPrefix and DisableNonWordSuffix

File: server/autolinker.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// NewAutoLinker create and initialize a AutoLinker
+func NewAutoLinker(link *Link) (*AutoLinker, error) {
+	if link == nil || len(link.Pattern) == 0 || len(link.Template) == 0 {
+		return nil, errors.New("Pattern or template was empty")
+	}
+
+	if !link.DisableNonWordPrefix {
+		link.Pattern = "(?P<MMDisableNonWordPrefix>^|\\s)" + link.Pattern
```
Comment (jfrerich) [inline server/autolinker.go]: s/MMDisableNonWordPrefix/DisableNonWordPrefix/ to maintain consistency with Field names in Link Struct

File: README.md (hunk 5 above)
Comment (jfrerich) [reply README.md]: Add documentation to README for the following:
- DisableNonWordPrefix
- DisableNonWordSuffix
- Scope

File: server/autolinker_test.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAutolink(t *testing.T) {
+	var tests = []struct {
+		Name            string
+		Link            *Link
+		inputMessage    string
+		expectedMessage string
+	}{
```
Comment (jfrerich) [inline server/autolinker_test.go]: Add Scope testing?

**Comment by jfrerich** [comment]
I loaded the plugin in my MM instance and ran some basic testing and also ran `go test`.  LGTM with some minor enhancement requests.

**Pull request closed without merging**